doc/widget_server.md

```diff
@@ -0,0 +1,417 @@
+← {response: "ok", seq_num: 1, record: {..., widget: {line: 25, column: 0, id: 51}}}
+→ {command: "get_widget", seq_num: 2, file_name: "my_file.lean", line: 25, column: 0, id: 51}
+← {response: "ok", seq_num: 2, widget: {line: 25, column: 0, id: 51, html: {...}}}
+
+## Rendering a widget
+
+The recommended UI framework for use with rendering widgets is React, however this is not required.
+A reference implementation for React can be found [here](https://github.com/leanprover/vscode-lean/blob/master/infoview/widget.tsx).
+
+After receiving a `GetWidgetResponse` object `r`, the DOM information is stored in `r.widget.html`. This has type `WidgetComponent`, whose schema can be viewed below:
+
+
+```ts
+export interface WidgetEventHandler {
+    /** handler id */
+    h: number;
+    /** route */
+    r: number[];
+}
+
+export interface WidgetElement {
+    /** tag */
+    t: string;
+    /** children */
+    c: WidgetHtml[];
+    /** attributes */
+    a?: { [k: string]: any };
+    /** events */
+    e: {
+        'onClick'?: WidgetEventHandler;
+        'onMouseEnter'?: WidgetEventHandler;
+        'onMouseLeave'?: WidgetEventHandler;
+    };
+    /** tooltip */
+    tt?: WidgetHtml;
+}
+
+export interface WidgetComponent {
+    /** children */
+    c: WidgetHtml[];
+}
+
+export type WidgetHtml =
+    | WidgetComponent
+    | string
+    | WidgetElement
+    | null;
+```
+
+### Rendering `WidgetElement`
+
+Ignoring tooltips for now, each `WidgetElement` corresponds to a DOM element.
+
+- `"t"` is the tag of the element.
+- `"a"` are the non-event attributes of the element.
+- `"e"` are the event handlers that should be attached to the element.
+- `"c"` are the children of the element.
+- `"tt"` is an additional tooltip element to be discussed below.
+
+For example:
+
+```json
+{
+  "t": "button",
+  "a": {
+      "style": {
+          "color": "red",
+          "padding": "5em",
+          "textAlign": "left"
+      },
+      "className": "b--blue ba",
+  },
+  "e": {
+      "onClick": { "h": 0, "r": [ 51 ] }
+  },
+  "c": ["hello", "world"]
+}
+```
+
+Converts to the React element:
+
+```tsx
+<button
+    style={{color:"red", padding:"5em", textAlign:"left"}}
+    className="b--blue ba"
+    onClick={() => handle({h:0, r:[51]}, 'onClick')}>
+  helloworld
+</button>
+```
+
+
+#### Attributes
+The attributes field of `WidgetElement` takes the form of a JSON object of keys and values.
+Either both the key and value are a string, or a special case is `"style"`, whose value is a JSON object of string key/value pairs.
+Events attached to the element are placed in their own `e` field.
+Note: the keys of attributes and styles follow the React convention; for example `"textAlign"` and `"onClick"` instead of `"text-align"` and `"on-click"`. Additionally, `"className"` is used instead of `"class"`. So if you are not implementing with React, you will need to convert all of these keynames back to their HTML-native forms.
```

Comment:
```suggestion

Note: the keys of attributes and styles follow the React convention; for example `"textAlign"` and `"onClick"` instead of `"text-align"` and `"on-click"`. Additionally, `"className"` is used instead of `"class"`. So if you are not implementing with React, you will need to convert all of these keynames back to their HTML-native forms.
```